fs_attachment/tests/test_fs_attachment.py

```diff
@@ -30,6 +30,38 @@ def test_create_attachment_explicit_location(self):
             f.write(b"new")
         self.assertEqual(attachment.raw, b"new")
 
+    def test_create_attachment_with_meaningful_name(self):
+        """In this test we use a backend with 'optimizes_directory_path',
+        which rewrites the filename to be a meaningful name.
+        We ensure that the rewritten path is consistently used,
+        meaning we can read the file after.
+        """
+        content = b"This is a test attachment"
+        attachment = (
+            self.env["ir.attachment"]
+            .with_context(
+                storage_location=self.backend_optimized.code,
+                force_storage_key="test.txt",
+            )
+            .create({"name": "test.txt", "raw": content})
+        )
+        # the expected store_fname is made of the storage code,
+        # a random middle part, and the filename
+        # example: tmp_opt://te/st/test-198-0.txt
+        # The storage root is NOT part of the store_fname
+        self.assertFalse("tmp/" in attachment.store_fname)
+
+        # we can read the file, so storage finds it correctly
+        with attachment.open("rb") as f:
+            self.assertEqual(f.read(), content)
+
+        new_content = b"new content"
+        with attachment.open("wb") as f:
+            f.write(new_content)
```

Comment:
IMO, after the 'write' you should also check the store_fname to be sure that the new file is till into an optimized directory path...